File: app/Laplace.Core/Core/SqlCatalog.cs
using System.Runtime.InteropServices;

namespace Laplace.Engine.Core;

public sealed record NativeSqlQuery(string Name, string Text, string[] ParameterTypes);

/// <summary>Immutable SQL and parameter contracts marshalled from the native catalog.</summary>
public static partial class SqlCatalog
{
    [LibraryImport("laplace_core", EntryPoint = "laplace_sql_query_text", StringMarshalling = StringMarshalling.Utf8)]
    private static partial IntPtr TextPointer(string name);

    [LibraryImport("laplace_core", EntryPoint = "laplace_sql_query_parameters", StringMarshalling = StringMarshalling.Utf8)]
    private static partial IntPtr ParametersPointer(string name);

    public static NativeSqlQuery Get(string name)
    {
        var text = Marshal.PtrToStringUTF8(TextPointer(name))
            ?? throw new ArgumentException($"Unknown native SQL query '{name}'.", nameof(name));
        var parameters = Marshal.PtrToStringUTF8(ParametersPointer(name))
            ?? throw new InvalidOperationException($"Missing native SQL parameter contract for '{name}'.");
        return new(name, text, parameters.Length == 0 ? [] : parameters.Split(','));
    }
}

File: engine/core/src/sql_catalog.c
#include "laplace/core/sql_catalog.h"
#include <stddef.h>
#include <string.h>

typedef struct { const char *name, *parameters, *text; } SqlQuery;
#define SQL_QUERY(name, parameters, text) {name, parameters, text},
static const SqlQuery queries[] = {
#include "sql_catalog.def"
};
#undef SQL_QUERY

static const SqlQuery *lookup(const char *name)
{
    if (name == NULL) return NULL;
    for (size_t i = 0; i < sizeof(queries) / sizeof(queries[0]); ++i)
        if (strcmp(name, queries[i].name) == 0) return &queries[i];
    return NULL;
}
const char *laplace_sql_query_text(const char *name)
{
    const SqlQuery *q = lookup(name);
    return q ? q->text : NULL;
}
const char *laplace_sql_query_parameters(const char *name)
{
    const SqlQuery *q = lookup(name);
    return q ? q->parameters : NULL;
}
